tests/unit/test_ansible_inventory.py

```diff
@@ -0,0 +1,77 @@
+from mrack.outputs.ansible_inventory import AnsibleInventoryOutput, get_group
+
+from .mock_data import (
+    common_inventory_layout,
+    create_metadata,
+    get_db_from_metadata,
+    provisioning_config,
+)
+
+
+def ensure_all_groups_present(metadata, inventory):
+    """
+    Ensure that all groups defined in metadata hosts objects are present.
+
+    And contain the host
+    """
+    for domain in metadata["domains"]:
+        for meta_host in domain["hosts"]:
+            for groupname in meta_host.get("groups"):
+                group = get_group(inventory, groupname)
+                assert group, "All defined groups in host must be in inventory"
+                assert "hosts" in group, "Group must contain hosts dict"
+                hosts = group["hosts"]
+                assert meta_host["name"] in hosts, "Group must contain the host"
+
+
+def ensure_hosts_in_all_group(db, inventory):
+    """Ensure that group 'all' contains all hosts from DB."""
+    all_group = inventory["all"]
+    hosts = all_group["hosts"]
+    db_hosts = db.hosts
+
+    assert len(db_hosts) > 0, "Make sure we do not work on empty data set"
+
+    required_attrs = [
+        "ansible_host",
+        "ansible_python_interpreter",
+        "ansible_user",
+        "meta_domain",
+        "meta_fqdn",
+        "meta_os",
+        "meta_ip",
+        "meta_provider_id",
+        "meta_role",
+    ]
+
+    for name, dbhost in db_hosts.items():
+        assert dbhost.name in hosts, "All hosts must be present in inventory"
+
+        invhost = hosts[dbhost.name]
+
+        assert dbhost.ip == invhost["meta_ip"], "IP from DB in inventory"
+        for attr in required_attrs:
+            assert attr in required_attrs, "All required attrs are in host definition"
+
+
+class TestAnsibleInventory:
+    def setup(self):
+        self.metadata = create_metadata(ipaservers=1, ipaclients=1, ads=1)
+        self.db = get_db_from_metadata(self.metadata)
+
+    def test_common_layout(self):
+
+        layout = common_inventory_layout()
+        config = provisioning_config(layout)
+
+        ans_inv = AnsibleInventoryOutput(config, self.db, self.metadata)
+
+        inventory = ans_inv.create_inventory()
```

Comment:
You can move this to a fixture called `inventory`; (You can do the same to `metadata` and `db`) Pass them to `test_common_layout` and move the `ensure_*` functions to become methods.

Fixtures are the best practice to setup and teardown scenarios.

Reply:
Amended and several tests added (through parametrize).